### ISSUE TITLE
how to disable a single comarisonOperatorUsage

### ISSUE THREAD

**Issue by adminfriso** [squizlabs/PHP_CodeSniffer#2238]
Hello,
I am working on a custom sniff for my (shared) projects. I would lik to allow the T_BOOLEAN_NOT as valid operator on the Squiz.Operators.ComparisonOperatorUsage.NotAllowed rule.
right now the only way (I could find)  to exclude this is to exclude the complete rule, but then you will also lose the other checks.

    <rule ref="Squiz.Operators.ComparisonOperatorUsage">
        <exclude name="Squiz.Operators.ComparisonOperatorUsage.NotAllowed"/>
    </rule>

is there a way to only exclude this single rule?

**Comment by dingo-d** [comment]
You can silence it, that way you won't be including the entire ruleset

```xml
<rule ref="Squiz.Operators.ComparisonOperatorUsage.NotAllowed">
  <severity>0</severity>
</rule>
```

**Comment by adminfriso** [comment]
But then i do also silence the check for the other valid and invald OPS

**Comment by adminfriso** [comment]
for example
 61 | ERROR | Operator ! prohibited; use === FALSE instead // Do not want to check for this
 66 | ERROR | Operator != prohibited; use !== instead // Do want to check for this

**Comment by dingo-d** [comment]
Oh, I see. You could remove this one, and create your own that is modified to your liking (basically fork the current one and make modifications that fit your use case).

**Comment by adminfriso** [comment]
how do you mean? and or is there a way to do it in the current package?

**Comment by dingo-d** [comment]
I'm assuming that you are just using `phpcs.xml.dist` in your project and cherrypicking the rules you want to (the usual way).

What I did was I [created a standard](https://github.com/infinum/coding-standards-wp) for the company I work for, and I added a custom sniff. You could do something like it, define your own sniff and it will be included in your ruleset.

Or you could fork the phpcs and change the sniff in question and then load your fork in the composer. This is a bit more work as you need to keep track of the changes in the upstream branch.

**Comment by jrfnl** [comment]
@adminfriso Just a heads-up: The simplest way to create your own version of this sniff would be to just extend the PHPCS sniff and overload the `$invalidOps` property. That way you are not infringing on the copyright of the code.

Unfortunately that property appears to be `private`, but a PR to this repo to change the property from `private` to `protected` should fix that.
I'm also wondering why the property is set to `static`, A search of the repo and tracing through the file history should be able to explain that. If there's no good reason, the `static` should probably be removed as well, making overloading the `$invalidOps` property even easier.

For an example of how to extend/overload a PHPCS sniff in your own external standard, you could have a look at this WPCS sniff which does something similar for another sniff: https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/blob/develop/WordPress/Sniffs/WhiteSpace/OperatorSpacingSniff.php

**Comment by jrfnl** [comment]
@gsherwood I'd favour more modular errorcodes, but that change would be a BC-break which would have to wait for the 4.0 version, wouldn't it ?

**Comment by gsherwood** [comment]
> but that change would be a BC-break which would have to wait for the 4.0 version, wouldn't it ?

It could technically be in a minor release given the rules I've set, but I don't think I'd push this in a minor given the sniff has a second error code.

If it had a single code, you could still mute the all the new codes by excluding the sniff. But if you are currently muting just `Squiz.Operators.ComparisonOperatorUsage.NotAllowed` and you want the other code, then you actually have a worse time with this change. You can't mute `Squiz.Operators.ComparisonOperatorUsage.*NotAllowed`.

**Comment by adminfriso** [comment]
You could also add idd in recursion so 
`Squiz.Operators.ComparisonOperatorUsage.NotAllowed.Method`

**Comment by VincentLanglet** [comment]
@gsherwood @jrfnl I'm interested about a sniff only checking '==' and '!=' are not used. And this check should be done in controle structure but also in assignement (`$foo = true == 1`).
I allow the boolean not operator and implicit true comparison.

Are you interested with a PR ? Should I add option to this sniff or create a new sniff ?

**Comment by gsherwood** [comment]
> Should I add option to this sniff or create a new sniff ?

Feels like a different sniff.

Ideally, this sniff is deprecated and 2 new sniffs are created; one to check for banned comparison operators, and one to check for implicit true comparisons.

A new sniff to ban some comparison operators could then either use sniff properties to let developers decide which operators to ban, and/or different sniff codes to specific messages could be muted. But I don't think I'd accept a sniff that just hard-codes a ban on `==` and `!=` as that would probably just add to the problem this issue is describing.

**Comment by VincentLanglet** [comment]
> A new sniff to ban some comparison operators could then either use sniff properties to let developers decide which operators to ban, and/or different sniff codes to specific messages could be muted. But I don't think I'd accept a sniff that just hard-codes a ban on `==` and `!=` as that would probably just add to the problem this issue is describing.

Is this draft could be ok ? https://github.com/squizlabs/PHP_CodeSniffer/pull/2803

I made the operator property public.
I'm not sure how having good different sniff code since, the token name are not always words (for example, "==" and "!=").

I'll maje the implicit true comparisons check after

**Comment by VincentLanglet** [comment]
> But I don't think I'd accept a sniff that just hard-codes a ban on `==` and `!=` as that would probably just add to the problem this issue is describing.

@gsherwood Like said here https://github.com/squizlabs/PHP_CodeSniffer/pull/2803#issuecomment-572321462, I don't think a sniff that hard-code a ban on `==` and `!=`, would add the same issue since we can consider this as a *minimal sniff*.
With a sniff for the `!` operator and a third sniff for the implicit `true` we have 3 minimal sniff. The developer can then have exactly the checks he wants by taking the sniff he need.

In the same way there is a sniff that hard-code a ban on `and` and `or`.

What do you think ?

**Comment by glen-84** [comment]
I needed this as well, so I wrote [a simple sniff](https://github.com/glen-84/phpcs-standard/blob/964967947b0e71a5d8b22ff8b0d8d6316b01f20b/Glen84/Sniffs/Operators/ComparisonOperatorUsageSniff.php) years ago.

**Comment by jrfnl** [comment]
Similar to: [`Wordpress.PHP.StrictComparisons`](https://github.com/WordPress/WordPress-Coding-Standards/blob/develop/WordPress/Sniffs/PHP/StrictComparisonsSniff.php)
